File: 64-minimum-path-sum/minimum-path-sum.py
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        # grid[1][1]+=grid[0][0]+min(grid[1][0],grid[0][1])
        n=len(grid)
        m=len(grid[0])
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if i==0 and j==0:
                    continue
                if i==0 and j!=0:
                    grid[i][j]+=grid[i][j-1]
                elif j==0 and i!=0:
                    grid[i][j]+=grid[i-1][j]
                else :
                    grid[i][j]+=min(grid[i][j-1],grid[i-1][j])
        return grid[n-1][m-1]
        
       

File: 64-minimum-path-sum/minimum-path-sum.cpp
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int n=grid.size();
        int m=grid[0].size();
        vector<int> prevRow(m,-1);

        for(int i=0;i<n;++i)
        {   vector<int> cur(m,-1);
            for(int j=0;j<m;++j)
            {
                if (i==0 && j==0) cur[j]=grid[i][j];
                else{
                    int up=INT_MAX,left=INT_MAX;
                    if(i>0) up=prevRow[j] + grid[i][j];
                    if(j>0) left=cur[j-1]+grid[i][j];
                    cur[j]=min(up,left);
                }
            }
            prevRow=cur;
        }return prevRow[m-1];
    }
};